src/types/profile.js

```diff
@@ -15,17 +15,45 @@ export type IndexIntoStringTable = number;
 export type IndexIntoFuncTable = number;
 export type IndexIntoResourceTable = number;
 export type IndexIntoLibs = number;
-export type categoryBitMask = number;
+export type IndexIntoCategoryList = number;
 export type resourceTypeEnum = number;
 export type ThreadIndex = number;
 
 /**
- * The stack table is the minimal representation of a call stack. Each stack entry
- * consists of the frame at the top of the stack, and the prefix for the stack that
- * came before it. Stacks can be shared between samples.
+ * The stack table stores the tree of stack nodes of a thread.
+ * The shape of the tree is encoded in the prefix column: Root stack nodes have
+ * null as their prefix, and every non-root stack has the stack index of its
+ * "caller" / "parent" as its prefix.
+ * Every stack node also has a frame and a category.
+ * A "call stack" is a list of frames. Every stack index in the stack table
+ * represents such a call stack; the "list of frames" is obtained by walking
+ * the path in the tree from the root to the given stack node.
+ *
+ * Stacks are used in the thread's samples; each sample refers to a stack index.
+ * Stacks can be shared between samples.
+ *
+ * With this representation, every sample only needs to store a single integer
+ * to identify the sample's stack.
+ * We take advantage of the fact that many call stacks in the profile have a
+ * shared prefix; storing these stacks as a tree saves a lot of space compared
+ * to storing them as actual lists of frames.
+ *
+ * The category of a stack node is always non-null and is derived from a stack's
+ * frame and its prefix. Frames can have null categories, stacks cannot. If a
+ * stack's frame has a null category, the stack inherits the category of its
+ * prefix stack. Root stacks whose frame has a null stack have their category
+ * set to the "default category". (The default category is currently defined as
+ * the category in the profile's category list whose color is "grey", and such
+ * a category is required to be present.)
+ *
+ * You could argue that the stack table's category column is derived data and as
+ * such doesn't need to be stored in the profile itself. This is true, but
+ * storing this information in the stack table makes it a lot easier to carry
+ * it through various transforms that we apply to threads.
  */
 export type StackTable = {
   frame: IndexIntoFrameTable[],
+  category: IndexIntoCategoryList[],
```

Comment:
Maybe a newbie question: why are we using normal arrays here, but Int32Arrays in the CallNodeTable ?

Reply:
No good reason. I think when I created the call node table, I had to work really hard to get it to perform well, so I tried to squeeze out performance in every corner that I could find. But I don't think I've ever properly benchmarked the typed array solution versus the regular array solution.

It's possible that the performance of the stackTable would also benefit from typed arrays.

There's one difference between the callNodeTable and the tables in the profile that makes it easier to use typed arrays in the callNodeTable: The callNodeTable is never stored inside the profile - it's just external (derived) information that's used for display. So we don't need to worry about serializing it. But with data in the profile, we need a way of serializing it to JSON / deserializing it from JSON. We already have some custom code for the stringTable, so it wouldn't be too hard to add more custom serialization code for the other tables, but I'm not sure if it's worth doing.